pkg/xds/context/mesh_snapshot.go

```diff
@@ -20,46 +20,66 @@ import (
 
 var meshCacheLog = core.Log.WithName("mesh-cache")
 
-// meshSnapshot represents all resources that belong to Mesh and allows to calculate hash.
+// MeshSnapshot represents all resources that belong to Mesh and allows to calculate hash.
 // Calculating and comparing hashes is much faster than call 'equal' for xDS resources. So
 // meshSnapshot reduces costs on reconciling Envoy config when resources in the store are
 // not changing
-type meshSnapshot struct {
+type MeshSnapshot struct {
 	mesh      *core_mesh.MeshResource
 	resources map[core_model.ResourceType]core_model.ResourceList
 	ipFunc    lookup.LookupIPFunc
 }
 
-func GetMeshSnapshot(ctx context.Context, meshName string, rm manager.ReadOnlyResourceManager, types []core_model.ResourceType, ipFunc lookup.LookupIPFunc) (*meshSnapshot, error) {
-	snapshot := &meshSnapshot{
+type MeshSnapshotBuilder interface {
+	Build(ctx context.Context, meshName string) (*MeshSnapshot, error)
+}
+
+func NewMeshSnapshotBuilder(rm manager.ReadOnlyResourceManager, types []core_model.ResourceType, ipFunc lookup.LookupIPFunc) MeshSnapshotBuilder {
+	return &meshSnapshotBuilder{
+		rm:     rm,
+		types:  types,
+		ipFunc: ipFunc,
+	}
+}
+
+type meshSnapshotBuilder struct {
+	rm     manager.ReadOnlyResourceManager
+	types  []core_model.ResourceType
+	ipFunc lookup.LookupIPFunc
```

Comment:
It seems like `ipFunc` is not used

Reply:
it is passed to the MeshSnapshot that later relies on it when counting hash. It's kind of weird, I've got an idea how to refactor it.